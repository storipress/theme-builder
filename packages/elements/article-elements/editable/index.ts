import { createPlaceholder } from './placeholder'

export { default as Article } from './article.vue'
export { default as ArticleBlock } from './article.vue'
export { default as ArticleContent } from './article-content.vue'
export { default as AuthorList } from './author-list.vue'
export { default as AuthorName } from './author-name.vue'
export { Authors } from './authors'
export { default as ColorArea } from './color-area.vue'
export { default as Content } from './content.vue'
export { default as Date } from './date.vue'
export { default as Description } from './description.vue'
export { default as Desk } from './desk.vue'
export { default as Header } from './header.vue'
export { default as HeaderBlock } from './header.vue'
export { HeadlineCaption, HeadlineImage } from './headline-image'
export { default as HeadlineWrapper } from './headline-wrapper.vue'
export { default as Title } from './title.vue'
export { default as TitleElement } from './title.vue'

export const Header1 = createPlaceholder('Header1', 'h1')
export const Header2 = createPlaceholder('Header2', 'h2')
export const Paragraph = createPlaceholder('Paragraph', 'p')
export const Blockquote = createPlaceholder('Blockquote', 'blockquote')
