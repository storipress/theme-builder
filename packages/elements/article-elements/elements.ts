import { createSwitch } from './switch'

export { default as LinkElement } from '../common/link-element.vue'
export { default as ResponsiveImage } from '../common/responsive-image.vue'
export { default as Site } from '../common/site.vue'
export { default as ArticleInfo } from './article-info.vue'
export { default as Logo } from './logo.vue'
export { default as Icon } from 'shared/components/icon.vue'

export const Article = createSwitch('Article')
export const ArticleBlock = createSwitch('ArticleBlock')
export const ArticleContent = createSwitch('ArticleContent')
export const AuthorList = createSwitch('AuthorList')
export const AuthorName = createSwitch('AuthorName')
export const Authors = createSwitch('Authors')
export const Blockquote = createSwitch('Blockquote')
export const ColorArea = createSwitch('ColorArea')
export const Content = createSwitch('Content')
export const Date = createSwitch('Date')
export const Description = createSwitch('Description')
export const Desk = createSwitch('Desk')
export const Header = createSwitch('Header')
export const HeaderBlock = createSwitch('HeaderBlock')
export const Header1 = createSwitch('Header1')
export const Header2 = createSwitch('Header2')
export const HeadlineCaption = createSwitch('HeadlineCaption')
export const HeadlineImage = createSwitch('HeadlineImage')
export const HeadlineWrapper = createSwitch('HeadlineWrapper')
export const Paragraph = createSwitch('Paragraph')
export const Title = createSwitch('Title')
export const TitleElement = createSwitch('TitleElement')
