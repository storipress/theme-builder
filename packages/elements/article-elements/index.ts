export * from '../common/subscribe'
export { convertNameToPath } from '../utils/article'
export * from './elements'
export type { ArticleInjected, Events } from './inject'
export { Template } from './template'
