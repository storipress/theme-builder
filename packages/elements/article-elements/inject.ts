import type { VueConstructor } from 'vue'
import type { CommonInjected } from '../common/common-element'
import { Evt } from 'evt'
import { constant, noop } from 'lodash'
import Vue from 'vue'

import { inject, ref } from 'vue-demi'

type RegisterFunction = (path: string[], props: Record<string, unknown>) => void

type Maybe<T> = T | null

export type Events = [event: 'focus', name: string] | [event: 'preview', value: void]

interface ElementInfo {
  name: string
  display?: string
  path: string[]
}

interface ElementState {
  hover: Maybe<ElementInfo>
  selected: Maybe<ElementInfo>
}

export interface Focus {
  x: number
  y: number
}

export interface Author {
  name: string
  url: string
}

export interface FocalInfo {
  original: Focus
  url: string
}

export interface SetData<T> {
  path: string[]
  data: T
  skipHistory?: boolean
  breakpoint?: string
}

// TODO: split editable & stylable attributes?
export interface ArticleInjected extends CommonInjected {
  section: ElementState
  elements: Record<string, string>
  selectable: boolean
  editable: boolean

  title: string
  blurb: string
  desk: string
  authors: Author[]
  date: Date
  headlineURL: string | null
  headlineFocus: Focus
  headlineAlt: string
  headlineCaption: string
  editor: object | null
  profile: object
  bus: Evt<Events>

  createImageURL: (url: string, edits?: Record<string, string>) => string
  uploadImage: (file: File) => Promise<string>

  readStyle: (path: readonly string[]) => Record<string, unknown>
  setElementStyle: (payload: SetData<Record<string, unknown>>) => void
  setSectionHover: (info: ElementInfo | null) => void
  setSectionSelect: (info: ElementInfo | null) => void
  addColor: (info: { path: string[]; style: Record<string, unknown> }) => void
  registerElementDefault: RegisterFunction
}

export interface Injected {
  $element: ArticleInjected
}

type withInjected = VueConstructor<Vue & Injected>

export const INJECTED_DEFAULT: ArticleInjected = {
  section: {
    hover: null,
    selected: null,
  },
  elements: {},
  selectable: true,
  editable: false,
  scale: 1,

  site: { name: '' },
  logo: '',

  title: '',
  blurb: '',
  desk: '',
  authors: [],
  date: new Date(),
  headlineURL: null,
  headlineFocus: { x: 0, y: 0 },
  headlineAlt: '',
  headlineCaption: '',
  editor: null,
  profile: {},
  bus: Evt.create<Events>(),

  createImageURL: constant(''),
  uploadImage: constant(Promise.resolve('')),

  readStyle: constant({}),
  setElementStyle: noop,
  setSectionHover: noop,
  setSectionSelect: noop,
  addColor: noop,
  registerElementDefault: noop,
}

export const BaseElement = (Vue as withInjected).extend({
  inject: {
    $element: {
      default: constant(INJECTED_DEFAULT),
    },
  },
})

export function useArticleElement() {
  // wrap in ref to prevent auto unwrap
  return ref(inject<ArticleInjected>('$element', INJECTED_DEFAULT))
}
