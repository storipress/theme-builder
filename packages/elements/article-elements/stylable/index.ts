export { default as Article } from './article.vue'
export { default as ArticleBlock } from './article.vue'
export { default as ArticleContent } from './article-content.vue'
export { default as AuthorList } from './author-list.vue'
export { default as AuthorName } from './author-name.vue'
export { Authors } from './authors'
export { default as Blockquote } from './blockquote.vue'
export { default as ColorArea } from './color-area.vue'
export { default as Content } from './content.vue'
export { default as Date } from './date.vue'
export { default as Description } from './description.vue'
export { default as Desk } from './desk.vue'
export { default as Header } from './header.vue'
export { default as HeaderBlock } from './header.vue'
export { default as Header1 } from './header1.vue'
export { default as Header2 } from './header2.vue'
export { default as HeadlineCaption } from './headline-caption.vue'
export { default as HeadlineImage } from './headline-image.vue'
export { default as HeadlineWrapper } from './headline-wrapper.vue'
export { default as Paragraph } from './paragraph.vue'
export { default as Title } from './title.vue'
export { default as TitleElement } from './title.vue'
