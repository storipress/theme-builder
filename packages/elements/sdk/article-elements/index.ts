export * from '../../common/subscribe'
export { _validateData } from '../common'
export * from './elements'
export type { ArticleInjected } from './inject'
export { Template } from './template'
