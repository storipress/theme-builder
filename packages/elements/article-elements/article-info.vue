<script lang="ts">
import { computed, defineComponent, readonly } from 'vue-demi'

import { useArticleElement } from './inject'

export default defineComponent({
  setup() {
    const element = useArticleElement()

    return {
      article: computed(() => {
        const { title, blurb, headlineCaption, desk, authors, date, headlineURL, headlineFocus, headlineAlt } =
          element.value

        return {
          title,
          blurb,
          headlineCaption,
          desk,
          deskUrl: '#',
          authors,
          date,
          headlineURL,
          headlineFocus,
          headlineAlt,

          headline: headlineURL,
        }
      }),
      socials: readonly({
        facebook: '#',
        twitter: '#',
        email: '#',
      }),
    }
  },
})
</script>

<template>
  <div>
    <slot :article="article" :socials="socials" />
  </div>
</template>
