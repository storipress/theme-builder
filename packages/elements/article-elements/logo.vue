<script lang="ts">
import { ResponsiveImage } from '@storipress/elements/common'
import { computed, defineComponent } from 'vue-demi'

import { useArticleElement } from './inject'

export default defineComponent({
  components: { ResponsiveImage },

  setup() {
    const $element = useArticleElement()

    const logo = computed(() => $element.value.logo)

    return {
      logo,
    }
  },
})
</script>

<template>
  <ResponsiveImage v-if="logo" :src="logo" alt="logo" />
</template>
