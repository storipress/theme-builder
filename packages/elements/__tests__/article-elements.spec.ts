import { describe, expect, it, vi } from 'vitest'
import * as siteVariant from '../article-elements'
import * as sdkVariant from '../sdk/article-elements'

vi.mock('@storipress/editor-core', () => ({
  Tiptap: { render: (h: any) => h() },
  RichInput: { render: (h: any) => h() },
}))

function isComponentLike(name: string): boolean {
  return name[0].toUpperCase() === name[0]
}

describe('article elements', () => {
  it('should have the same properties as the site variant', () => {
    const keys = Object.keys(siteVariant).filter((key) => isComponentLike(key))
    for (const key of keys) {
      expect(sdkVariant).toHaveProperty(key)
    }
    expect(keys).toMatchSnapshot()
  })
})
