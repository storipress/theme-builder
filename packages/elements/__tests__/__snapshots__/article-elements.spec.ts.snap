// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`article elements > should have the same properties as the site variant 1`] = `
[
  "SubscribeEmail",
  "SubscribeForm",
  "LinkElement",
  "ResponsiveImage",
  "Site",
  "ArticleInfo",
  "Logo",
  "Icon",
  "Article",
  "ArticleBlock",
  "ArticleContent",
  "AuthorList",
  "AuthorName",
  "Authors",
  "Blockquote",
  "ColorArea",
  "Content",
  "Date",
  "Description",
  "Desk",
  "Header",
  "HeaderBlock",
  "Header1",
  "Header2",
  "HeadlineCaption",
  "HeadlineImage",
  "HeadlineWrapper",
  "Paragraph",
  "Title",
  "TitleElement",
  "Template",
]
`;
